pkg/metadata/container.go

```diff
@@ -0,0 +1,183 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metadata
+
+import (
+	"encoding/json"
+
+	"github.com/kubernetes-incubator/cri-containerd/pkg/metadata/store"
+
+	"k8s.io/kubernetes/pkg/kubelet/api/v1alpha1/runtime"
+)
+
+// The code is very similar with sandbox.go, but there is no template support
+// in golang, we have to have similar files for different types.
+// TODO(random-liu): Figure out a way to simplify this.
+// TODO(random-liu): Handle versioning with the same mechanism with container.go
+
+// containerMetadataVersion  is current version of container metadata.
+const containerMetadataVersion = "v1" // nolint
+
+// versionedContainerMetadata is the internal versioned container metadata.
+// nolint
+type versionedContainerMetadata struct {
+	// Version indicates the version of the versioned container metadata.
+	Version string
+	ContainerMetadata
+}
+
+// ContainerMetadata is the unversioned container metadata.
+type ContainerMetadata struct {
+	// ID is the container id.
+	ID string
+	// Name is the container name.
+	Name string
+	// SandboxID is the sandbox id the container belongs to.
+	SandboxID string
+	// Config is the CRI container config.
+	Config *runtime.ContainerConfig
+	// ImageRef is the reference of image used by the container.
+	ImageRef string
+	// Pid is the init process id of the container.
+	Pid uint32
+	// CreatedAt is the created timestamp.
+	CreatedAt int64
+	// StartedAt is the started timestamp.
+	StartedAt int64
+	// FinishedAt is the finished timestamp.
+	FinishedAt int64
+	// ExitCode is the container exit code.
+	ExitCode int32
+	// CamelCase string explaining why container is in its current state.
+	Reason string
+	// Human-readable message indicating details about why container is in its
+	// current state.
+	Message string
+	// Removing indicates that the container is in removing state.
+	// In fact, this field doesn't need to be checkpointed.
+	// TODO(random-liu): Skip this during serialization when we put object
+	// into the store directly.
+	// TODO(random-liu): Reset this field to false during state recoverry.
+	Removing bool
```

Comment:
Offline discussed with @yujuhong.

`Removing` is mainly introduced to avoid race with concurrent remove/start operations (mainly start).

We don't need this for sandbox because sandbox only has `run` instead of `create` and `start`. There is no race between `run` and `remove`.